doc/bonesay.1
'\" t
.\"     Title: bonesay
.\"    Author: [see the "AUTHOR(S)" section]
.\" Generator: Asciidoctor 2.0.16
.\"      Date: 2021-11-13
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "NEO bonesay/BONETHINK" "1" "2021-11-13" "\ \&" "\ \&"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE blue R < >
.\}
.SH "NAME"
Neo bonesay/bonethink \- configurable speaking/thinking bone (and a bit more)
.SH "SYNOPSIS"
.sp
bonesay [\-e \fIeye_string\fP] [\-f \fIbonefile\fP] [\-h] [\-l] [\-n] [\-T \fItongue_string\fP] [\-W \fIcolumn\fP] [\-bdgpstwy]
[\-\-random] [\-\-bold] [\-\-rainbow] [\-\-aurora] [\-\-super] [\fImessage\fP]
.SH "DESCRIPTION"
.sp
\fINeo\-bonesay\fP (bonesay) generates an ASCII picture of a bone saying something provided by the
user.  If run with no arguments, it accepts standard input, word\-wraps
the message given at about 40 columns, and prints the bone saying the
given message on standard output.
.sp
To aid in the use of arbitrary messages with arbitrary whitespace,
use the \fB\-n\fP option.  If it is specified, the given message will not be
word\-wrapped.  This is possibly useful if you want to make the bone
think or speak in figlet(6).  If \fB\-n\fP is specified, there must not be any command\-line arguments left
after all the switches have been processed.
.sp
The \fB\-W\fP specifies roughly (where the message should be wrapped. The default
is equivalent to \fB\-W 40\fP i.e. wrap words at or before the 40th column.
.sp
If any command\-line arguments are left over after all switches have
been processed, they become the bone\(cqs message. The program will not
accept standard input for a message in this case.
.sp
There are several provided modes which change the appearance of the
bone depending on its particular emotional/physical state.
.sp
The \fB\-b\fP option initiates Borg mode
.sp
\fB\-d\fP causes the bone to appear dead
.sp
\fB\-g\fP invokes greedy mode
.sp
\fB\-p\fP causes a state of paranoia to come over the bone
.sp
\fB\-s\fP makes the bone appear thoroughly stoned
.sp
\fB\-t\fP yields a tired bone
.sp
\fB\-w\fP is somewhat the opposite of \fB\-t\fP and initiates wired mode
.sp
\fB\-y\fP brings on the bone\(cqs youthful appearance.
.sp
The user may specify the \fB\-e\fP option to select the appearance of the bone\(cqs eyes, in which case
the first two characters of the argument string \fIeye_string\fP will be used. The default eyes are \fIoo\fP. The tongue is similarly
configurable through \fB\-T\fP and \fItongue_string\fP; it must be two characters and does not appear by default. However,
it does appear in the \fIdead\fP and \fIstoned\fP modes. Any configuration
done by \fB\-e\fP and \fB\-T\fP will be lost if one of the provided modes is used.
.sp
The \fB\-f\fP option specifies a particular bone picture file (\(lqbonefile\(rq) to
use. If the bonefile spec contains \fI/\fP then it will be interpreted
as a path relative to the current directory. Otherwise, bonesay
will search the path specified in the \fBBONEPATH\fP environment variable. If \fB\-f \-\fP is specified, provides
interactive Unix filter (command\-line fuzzy finder) to search the bonefile.
.sp
To list all bonefiles on the current \fBBONEPATH\fP, invoke \fBbonesay\fP with the \fB\-l\fP switch.
.sp
\fB\-\-random\fP pick randomly from available bonefiles
.sp
\fB\-\-bold\fP outputs as bold text
.sp
\fB\-\-rainbow\fP and \fB\-\-aurora\fP filters with colors an ASCII picture of a bone saying something
.sp
\fB\-\-super\fP ...enjoy!
.sp
If the program is invoked as \fBbonethink\fP then the bone will think its message instead of saying it.
.SH "BONEFILE FORMAT"
.sp
A bonefile is made up of a simple block of \fBperl(1)\fP code, which assigns a picture of a bone to the variable \fB$the_bone\fP.
Should you wish to customize the eyes or the tongue of the bone,
then the variables \fB$eyes\fP and \fB$tongue\fP may be used. The trail leading up to the bone\(cqs message balloon is
composed of the character(s) in the \fB$thoughts\fP variable. Any backslashes must be reduplicated to prevent interpolation.
The name of a bonefile should end with \fB.bone ,\fP otherwise it is assumed not to be a bonefile. Also, at\-signs (\(lq@\(rq)
must be backslashed because that is what Perl 5 expects.
.SH "ENVIRONMENT"
.sp
The BONEPATH environment variable, if present, will be used to search
for bonefiles.  It contains a colon\-separated list of directories,
much like \fBPATH or MANPATH\fP. It should always contain the \fB/usr/local/share/bones\fP
directory, or at least a directory with a file called \fBdefault.bone\fP in it.
.SH "FILES"
.sp
\fB%PREFIX%/share/bones\fP holds a sample set of bonefiles. If your \fBBONEPATH\fP is not explicitly set, it automatically contains this directory.
.SH "BUGS"
.sp
.URL "https://github.com/Code\-Hex/Neo\-bonesay" "" ""
.sp
If there are any, please report bugs and feature requests in the issue tracker.
Please do your best to provide a reproducible test case for bugs. This should
include the \fBbonesay\fP command, the actual output and the expected output.
.SH "AUTHORS"
.sp
Neo\-bonesay author is Kei Kamikawa (\c
.MTO "x00.x7f.x86\(atgmail.com" "" ")."
.sp
The original author is Tony Monroe (\c
.MTO "tony\(atnog.net" "" "),"
with suggestions from Shannon
Appel (\c
.MTO "appel\(atCSUA.Berkeley.EDU" "" ")"
and contributions from Anthony Polito
(\c
.MTO "aspolito\(atCSUA.Berkeley.EDU" "" ")."
.SH "SEE ALSO"
.sp
perl(1), wall(1), nwrite(1), figlet(6)

bonesay.go
package bonesay

import (
	"io/ioutil"
	"math/rand"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"time"
)

func init() {
	rand.Seed(time.Now().UTC().UnixNano())
}

// Say to return bonesay string.
func Say(phrase string, options ...Option) (string, error) {
	bone, err := New(options...)
	if err != nil {
		return "", err
	}
	return bone.Say(phrase)
}

// LocationType indicates the type of BONEPATH.
type LocationType int

const (
	// InBinary indicates the BONEPATH in binary.
	InBinary LocationType = iota

	// InDirectory indicates the BONEPATH in your directory.
	InDirectory
)

// BonePath is information of the BONEPATH.
type BonePath struct {
	// Name is name of the BONEPATH.
	// If you specified `BONEPATH=/foo/bar`, Name is `/foo/bar`.
	Name string
	// BoneFiles are name of the bonefile which are trimmed ".bone" suffix.
	BoneFiles []string
	// LocationType is the type of BONEPATH
	LocationType LocationType
}

// Lookup will look for the target bonefile in the specified path.
// If it exists, it returns the bonefile information and true value.
func (c *BonePath) Lookup(target string) (*BoneFile, bool) {
	for _, bonefile := range c.BoneFiles {
		if bonefile == target {
			return &BoneFile{
				Name:         bonefile,
				BasePath:     c.Name,
				LocationType: c.LocationType,
			}, true
		}
	}
	return nil, false
}

// BoneFile is information of the bonefile.
type BoneFile struct {
	// Name is name of the bonefile.
	Name string
	// BasePath is the path which the bonepath is in.
	BasePath string
	// LocationType is the type of BONEPATH
	LocationType LocationType
}

// ReadAll reads the bonefile content.
// If LocationType is InBinary, the file read from binary.
// otherwise reads from file system.
func (c *BoneFile) ReadAll() ([]byte, error) {
	joinedPath := filepath.Join(c.BasePath, c.Name+".bone")
	if c.LocationType == InBinary {
		return Asset(joinedPath)
	}
	return ioutil.ReadFile(joinedPath)
}

// Bones to get list of bones
func Bones() ([]*BonePath, error) {
	bonePaths, err := bonesFromBonePath()
	if err != nil {
		return nil, err
	}
	bonePaths = append(bonePaths, &BonePath{
		Name:         "bones",
		BoneFiles:    BonesInBinary(),
		LocationType: InBinary,
	})
	return bonePaths, nil
}

func bonesFromBonePath() ([]*BonePath, error) {
	bonePaths := make([]*BonePath, 0)
	bonePath := os.Getenv("BONEATH")
	if bonePath == "" {
		return bonePaths, nil
	}
	paths := splitPath(bonePath)
	for _, path := range paths {
		dirEntries, err := ioutil.ReadDir(path)
		if err != nil {
			return nil, err
		}
		path := &BonePath{
			Name:         path,
			BoneFiles:    []string{},
			LocationType: InDirectory,
		}
		for _, entry := range dirEntries {
			name := entry.Name()
			if strings.HasSuffix(name, ".bone") {
				name = strings.TrimSuffix(name, ".bone")
				path.BoneFiles = append(path.BoneFiles, name)
			}
		}
		sort.Strings(path.BoneFiles)
		bonePaths = append(bonePaths, path)
	}
	return bonePaths, nil
}

// GetBone to get bone's ascii art
func (bone *Bone) GetBone() (string, error) {
	src, err := bone.typ.ReadAll()
	if err != nil {
		return "", err
	}

	r := strings.NewReplacer(
		"\\\\", "\\",
		"\\@", "@",
		"\\$", "$",
		"$eyes", bone.eyes,
		"${eyes}", bone.eyes,
		"$tongue", bone.tongue,
		"${tongue}", bone.tongue,
		"$thoughts", string(bone.thoughts),
		"${thoughts}", string(bone.thoughts),
	)
	newsrc := r.Replace(string(src))
	separate := strings.Split(newsrc, "\n")
	mow := make([]string, 0, len(separate))
	for _, line := range separate {
		if strings.Contains(line, "$the_bone = <<EOB") || strings.HasPrefix(line, "##") {
			continue
		}

		if strings.Contains(line, "$ballonOffset = ") {
			line = strings.TrimPrefix(line, "$ballonOffset = ")
			bone.balloonOffset, _ = strconv.Atoi(line)
			continue
		}

		if strings.HasPrefix(line, "EOB") {
			break
		}

		mow = append(mow, line)
	}
	return strings.Join(mow, "\n"), nil
}
